test/affinity_tests.cc

```diff
@@ -0,0 +1,323 @@
+	skip_unsupported_configurations();
+	test_utils::allow_max_log_level(detail::log_level::warn);
+	env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "1,2");
+	{ raii_test_runtime rt(1); }
+	CHECK(test_utils::log_contains_substring(detail::log_level::warn, "Hardcoded core ID count (2) does not match the number of threads to be pinned ("));
+}
+
+// SimSYCL has no backend submission thread support
+#if !CELERITY_SYCL_IS_SIMSYCL
+TEST_CASE_METHOD(test_utils::runtime_fixture, "runtime system claims to pin its threads as desired", "[affinity][runtime]") {
+	skip_unsupported_configurations();
+	auto test = [](const std::vector<uint32_t>& core_ids) {
+		if(!have_cores({core_ids.cbegin(), core_ids.cend()})) {
+			SKIP("Skipping test because not all needed cores are available");
+			return;
+		}
+		REQUIRE(core_ids.size() == 5);
+
+		{ raii_test_runtime rt(2); }
+
+		constexpr auto msg_template = "Affinity: pinned thread '{}' to core {}";
+		CHECK(test_utils::log_contains_substring(detail::log_level::debug, fmt::format(msg_template, "application", core_ids.at(0))));
+		CHECK(test_utils::log_contains_substring(detail::log_level::debug, fmt::format(msg_template, "scheduler", core_ids.at(1))));
+		CHECK(test_utils::log_contains_substring(detail::log_level::debug, fmt::format(msg_template, "executor", core_ids.at(2))));
+		CHECK(test_utils::log_contains_substring(detail::log_level::debug, fmt::format(msg_template, "backend_worker_0", core_ids.at(3))));
+		CHECK(test_utils::log_contains_substring(detail::log_level::debug, fmt::format(msg_template, "backend_worker_1", core_ids.at(4))));
+	};
+
+	SECTION("for auto") {
+		env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "auto");
+		test({1, 2, 3, 4, 5});
+	}
+	SECTION("for hardcoded") {
+		env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "2,3,7,6,6");
+		test({2, 3, 7, 6, 6});
+	}
+	SECTION("for from") {
+		env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "from:3");
+		test({3, 4, 5, 6, 7});
+	}
+}
+#endif // !CELERITY_SYCL_IS_SIMSYCL
+
+TEST_CASE_METHOD(test_utils::runtime_fixture, "when pinning disabled: rt warns on insufficient threads and does not pin", "[affinity][runtime]") {
+	skip_unsupported_configurations();
+	env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "false");
+	raii_affinity_masking mask({0});
+
+	celerity::queue q;
+
+	CHECK(test_utils::log_contains_substring(
+	    detail::log_level::warn, fmt::format("only {} logical cores are available to this process. It is recommended to assign at least", 1)));
+	CHECK_FALSE(test_utils::log_contains_substring(detail::log_level::debug, "Affinity"));
+}
+
+TEST_CASE_METHOD(test_utils::runtime_fixture, "the application thread is actually pinned when pinning is enabled", "[affinity][runtime]") {
+	skip_unsupported_configurations();
+	env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "auto");
+
+	const core_set process_mask = {3, 4, 5, 6, 7};
```

Comment:
Why do we need this specific set, starting from 3? Couldn't this test also run on a 6-core machine? Same below.